clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -21,11 +24,41 @@ using namespace mlir::cir;
 Address CIRGenFunction::buildCompoundStmtWithoutScope(const CompoundStmt &S,
                                                       bool getLast,
                                                       AggValueSlot slot) {
-  for (auto *CurStmt : S.body())
-    if (buildStmt(CurStmt, /*useCurrentScope=*/false).failed())
-      return Address::invalid();
+  const Stmt *ExprResult = S.getStmtExprResult();
+  assert((!getLast || (getLast && ExprResult)) &&
+         "If getLast is true then the CompoundStmt must have a StmtExprResult");
 
-  return Address::invalid();
+  Address retAlloca = Address::invalid();
+
+  for (auto *CurStmt : S.body()) {
+    if (getLast && ExprResult == CurStmt) {
+      while (!isa<Expr>(ExprResult)) {
+        if (const auto *LS = dyn_cast<LabelStmt>(ExprResult))
+          llvm_unreachable("labels are NYI");
+        else if (const auto *AS = dyn_cast<AttributedStmt>(ExprResult))
+          llvm_unreachable("statement attributes are NYI");
+        else
+          llvm_unreachable("Unknown value statement");
+      }
+
+      const Expr *E = cast<Expr>(ExprResult);
+      QualType exprTy = E->getType();
+      if (hasAggregateEvaluationKind(exprTy)) {
+        buildAggExpr(E, slot);
+      } else {
+        // We can't return an RValue here because there might be cleanups at
+        // the end of the StmtExpr.  Because of that, we have to emit the result
+        // here into a temporary alloca.
+        retAlloca = CreateMemTemp(exprTy, getLoc(E->getSourceRange()));
+        buildAnyExprToMem(E, retAlloca, Qualifiers(),
+                          /*IsInit*/ false);
```

Comment:
Maybe we should delegate this to the lowering stage.

The idea here is that we need to backup values before the end of the scope executes any cleanups (e.g. destructors) that might erase the return value, but this could be done when lowering a `cir.scope` with a return value.

Reply:
I'm afraid we need the semantics in CIR, because of lifetime checking.

I also have the feeling we could do a bit different here. Since the lexical scope tracks the yielded value, perhaps we the cleanup handling could insert cleanup code in the retbb *before* the tracked `yield` op? Have you given that some thought? No need to change the implementation just yet, can you add a testcase that invokes a destructor so that I can help evaluate what to do?

Reply:
> I'm afraid we need the semantics in CIR, because of lifetime checking.

Can't we infer the allocation of a variable from the scope's return? E.g., if the scope returns a value, this implies the value is allocated in the current scope. Regardless, this would be mostly an aesthetic improvement, no need to dwell on it too much.

> the cleanup handling could insert cleanup code in the retbb before the tracked yield op? Have you given that some thought?

Not sure I follow. Is the idea to insert a new "return basic block" right before the `cir.yield` to dump any cleanup calls (e.g. destructors) in this new block?

> testcase that invokes a destructor so that I can help evaluate what to do?

I believe destructors are currently broken. I'll see if I can patch them.

Reply:
> Can't we infer the allocation of a variable from the scope's return? E.g., if the scope returns a value, this implies the value is allocated in the current scope

Scopes usually yield rvalue's, and those not necessarily mean allocations.

> I believe destructors are currently broken. I'll see if I can patch them.

Cool, in the meantime let's move on with this bit, you can address them in another PR.